docs/core/diagnostics/diagnosticsource-diagnosticlistener.md

```diff
@@ -0,0 +1,343 @@
+
+`DiagnosticListener` implements `IObservable<KeyValuePair<string, object>>`. This means
+ on each callback we get a `KeyValuePair`. The key of this pair is the name of the event
+ and the value is the payload `object`. In the code above we simply log this information
+ to the console.
+
+It is important to keep track of subscriptions to the `DiagnosticListener`. In the above code the
+networkSubscription variable remembers this. If another `creation` is formed, one must
+unsubscribe the previous listener and subscribe to the new one.
+
+The `DiagnosticSource`/`DiagnosticListener` code is thread safe, but the
+callback code also needs to be thread safe. To ensure the callback code is thread safe, locks are used. It is possible to create two `DiagnosticListeners`
+with the same name at the same time. To avoid races updates of shared variables are performed under the protection of a lock.
+
+Once the above code is run, the next time a `Write()` is done on 'System.Net.Http' `DiagnosticListener`
+the information will be logged to the console.
+
+Subscriptions are independent of one another. As a result, other code
+can do exactly the same thing as the code above, and generate two 'pipes' of the logging
+information.
+
+#### Decoding Payloads
+
+The `KeyvaluePair` that is passed to the callback has the event name and payload, but the payload is typed simply as
+an `object`. There are two ways of getting more specific data:
+
+1. If the payload is a well known type (e.g. a `string`, or an `HttpMessageRequest`) then you can simply
+   cast the `object` to the expected type (using the `as` operator so as not to cause an exception if
+   you are wrong) and then access the fields. This is very efficient.
+
+2. Use reflection API. For example, assume the method below is present.
+
+```C#
+    /// Define a shortcut method that fetches a field of a particular name.
+    static class PropertyExtensions
+    {
+        static object GetProperty(this object _this, string propertyName)
+        {
+            return _this.GetType().GetTypeInfo().GetDeclaredProperty(propertyName)?.GetValue(_this);
+        }
+    }
+```
+
+The `listener.Subscribe()` call above could be replaced with the following code, to decode the payload more fully.
+
+```C#
+    networkSubscription = listener.Subscribe(delegate(KeyValuePair<string, object> evnt) {
+        var eventName = evnt.Key;
+        var payload = evnt.Value;
+        if (eventName == "RequestStart")
+        {
+            var url = payload.GetProperty("Url") as string;
+            var request = payload.GetProperty("Request");
+            Console.WriteLine("Got RequestStart with URL {0} and Request {1}", url, request);
+        }
+    });
+```
+
+Note that using reflection is relatively expensive. However, using reflection is the only
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
```

Comment:
```suggestion
making fast, specialized property fetchers using either `PropertyInfo.CreateDelegate` or
```

Reply:
I'm not seeing either of these two methods in the .NET API browser?

Reply:
Based on the context, I'm guessing it means doing something like
```csharp
PropertyInfo pi = ...;
pi.GetMethod.CreateDelegate(...)
```

Where `CreateDelegate` is referring to `MethodInfo.CreateDelegate`.

e.g., [its usage in `DiagnosticSourceEventSoruce`.](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/DiagnosticSourceEventSource.cs,1392)

Reply:
@noahfalk are you aware of why CreateDelegate and ReflectEmit are not listed in the PropertyInfo [page](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.propertyinfo?view=net-6.0)?

Reply:
As far as I know those methods never existed. Probably what Vance (the original author of this doc) meant was propInfo.GetMethod.CreateDelegate() as @josalem was suggesting.